File: go/01_basic_chat.go
package main

// Example 1: Basic Chat
// This sends a simple question to the AI model using Go's standard library.
// No external dependencies needed!

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
)

// Request structures - define the JSON we send to the API
type ChatRequest struct {
	// Model is the AI model identifier to use for the completion
	// Example: "meta-llama/llama-4-scout-17b-16e-instruct"
	Model string `json:"model"`

	// Messages is the conversation history (array of user/assistant messages)
	// Each message has a role ("user" or "assistant") and content (the text)
	Messages []Message `json:"messages"`

	// Temperature controls randomness in responses (0.0 to 2.0)
	// 0.0 = deterministic/focused, 1.0 = balanced, 2.0 = very creative/random
	// Optional field (omitempty means it won't be sent if zero value)
	Temperature float64 `json:"temperature,omitempty"`

	// MaxTokens limits the length of the AI's response
	// 1 token ≈ 0.75 words. Typical values: 100-2000
	// Optional field
	MaxTokens int `json:"max_tokens,omitempty"`
}

type Message struct {
	// Role identifies who sent this message
	// Valid values: "user" (your input), "assistant" (AI response), "system" (behavior instructions)
	Role string `json:"role"`

	// Content is the actual text of the message
	Content string `json:"content"`
}

// Response structures - define the JSON we receive from the API
type ChatResponse struct {
	// ID is a unique identifier for this API request
	// Example: "chatcmpl-123abc"
	ID string `json:"id"`

	// Object describes the type of response
	// Always "chat.completion" for chat completions
	Object string `json:"object"`

	// Created is the Unix timestamp (seconds since 1970) when the response was generated
	Created int64 `json:"created"`

	// Model is the actual model that was used (may differ from requested if fallback occurred)
	Model string `json:"model"`

	// Choices contains the AI's response(s) - usually just one choice
	Choices []Choice `json:"choices"`

	// Usage provides token consumption statistics for billing/monitoring
	Usage Usage `json:"usage"`
}

type Choice struct {
	// Index identifies which choice this is (0-based)
	// Multiple choices can be requested, but we typically use just one
	Index int `json:"index"`

	// Message contains the AI's actual response
	Message Message `json:"message"`

	// FinishReason indicates why the response ended
	// Values: "stop" (natural end), "length" (hit max_tokens), "content_filter" (safety)
	FinishReason string `json:"finish_reason"`
}

type Usage struct {
	// PromptTokens is the number of tokens in your input (question + context)
	// Used to calculate input costs
	PromptTokens int `json:"prompt_tokens"`

	// CompletionTokens is the number of tokens in the AI's response
	// Used to calculate output costs (usually more expensive than input)
	CompletionTokens int `json:"completion_tokens"`

	// TotalTokens is the sum of prompt + completion tokens
	// Total cost = (PromptTokens × input_price) + (CompletionTokens × output_price)
	TotalTokens int `json:"total_tokens"`
}

// MAIN FUNCTION OVERVIEW:
// =======================
// What this example demonstrates:
//   - Basic HTTP POST request to Groq AI API
//   - JSON serialization and deserialization in Go
//   - Error handling patterns in Go
//   - Working with environment variables
//
// What you'll learn:
//   - How to structure API requests with Go's standard library
//   - How to handle JSON data without external dependencies
//   - How to read API responses and extract information
//   - Go's idiomatic error handling approach
//
// Expected output:
//   - Full JSON response from the API (prettified)
//   - The AI's answer to your question
//   - Token usage statistics (prompt, completion, total)
//
func main() {
	// Step 1: Get API key from environment variable
	// The API key is sensitive information, so we store it as an environment variable
	// rather than hardcoding it in the source code
	apiKey := os.Getenv("DEMETERICS_API_KEY")

	// Check if API key exists before proceeding
	if apiKey == "" {
		fmt.Println("Error: DEMETERICS_API_KEY environment variable not set")
		fmt.Println("Run: export DEMETERICS_API_KEY='your_key_here'")
		os.Exit(1)
	}

	// Step 2: Create the request payload
	// We build a ChatRequest struct that will be converted to JSON
	// This struct contains all the information the API needs
	request := ChatRequest{
		Model: "meta-llama/llama-4-scout-17b-16e-instruct", // AI model to use
		Messages: []Message{
			{
				Role:    "user",                               // "user" means this is our question
				Content: "What is the capital of Switzerland?", // The actual question
			},
		},
		Temperature: 0.7, // Controls randomness (0.0 = deterministic, 2.0 = very random)
		MaxTokens:   100, // Maximum length of response
	}

	// Step 3: Convert struct to JSON
	// json.Marshal converts our Go struct into JSON bytes
	// This is necessary because HTTP requests send data as bytes
	jsonData, err := json.Marshal(request)

	// Always check for errors in Go!
	if err != nil {
		fmt.Printf("Error creating JSON: %v\n", err)
		os.Exit(1)
	}

	// Step 4: Create HTTP request
	// http.NewRequest creates a new HTTP request object
	// We need to specify the method (POST), URL, and body (our JSON data)
	req, err := http.NewRequest(
		"POST",                                            // HTTP method
		"https://api.demeterics.com/groq/v1/chat/completions", // API endpoint
		bytes.NewBuffer(jsonData),                         // Request body (our JSON data)
	)

	// Check if request creation succeeded
	if err != nil {
		fmt.Printf("Error creating request: %v\n", err)
		os.Exit(1)
	}

	// Step 5: Set HTTP headers
	// Headers tell the server what kind of data we're sending and how to authenticate
	req.Header.Set("Content-Type", "application/json")    // We're sending JSON
	req.Header.Set("Authorization", "Bearer "+apiKey)     // API key for authentication

	// Step 6: Send the HTTP request
	// Create an HTTP client and use it to send our request
	client := &http.Client{} // &http.Client{} creates a new HTTP client with default settings
	resp, err := client.Do(req)

	// Check if the request was sent successfully
	if err != nil {
		fmt.Printf("Error sending request: %v\n", err)
		os.Exit(1)
	}

	// defer ensures resp.Body.Close() runs when main() exits
	// This is critical to prevent resource leaks
	defer resp.Body.Close()

	// Step 7: Read the response body
	// io.ReadAll reads all bytes from the response body
	body, err := io.ReadAll(resp.Body)

	// Check if reading the response succeeded
	if err != nil {
		fmt.Printf("Error reading response: %v\n", err)
		os.Exit(1)
	}

	// Step 8: Parse the JSON response
	// Create a variable to hold the parsed response
	var response ChatResponse

	// json.Unmarshal converts JSON bytes back into a Go struct
	// &response is a pointer so Unmarshal can modify the variable
	err = json.Unmarshal(body, &response)

	// Check if JSON parsing succeeded
	if err != nil {
		fmt.Printf("Error parsing JSON: %v\n", err)
		fmt.Printf("Raw response: %s\n", string(body))
		os.Exit(1)
	}

	// Step 9: Display results
	// Now that we have the parsed response, we can extract and display the information

	// Display the full response in a pretty format
	fmt.Println("Full Response:")
	prettyJSON, _ := json.MarshalIndent(response, "", "  ") // MarshalIndent adds formatting
	fmt.Println(string(prettyJSON))

	// Display just the AI's answer
	fmt.Println("\nAI Answer:")
	fmt.Println(response.Choices[0].Message.Content) // Access the first (and usually only) choice

	// Display token usage statistics
	// Tokens are units of text (roughly 4 characters per token)
	fmt.Println("\nToken Usage:")
	fmt.Printf("  Prompt: %d\n", response.Usage.PromptTokens)        // Tokens in our question
	fmt.Printf("  Response: %d\n", response.Usage.CompletionTokens)  // Tokens in AI's answer
	fmt.Printf("  Total: %d\n", response.Usage.TotalTokens)          // Sum of both
}

// Go concepts explained:
//
// package main
//   Every Go program starts with a package declaration
//   "main" package is special - it's the entry point
//
// import (...)
//   Import packages (libraries) we need
//   Standard library packages don't need installation
//
// type MyStruct struct {...}
//   Defines a custom data type (like a class in other languages)
//   Fields must start with capital letter to be exported (public)
//
// `json:"field_name"`
//   Struct tags - tell json.Marshal/Unmarshal how to handle fields
//   "omitempty" = don't include if zero value (0, "", nil, etc.)
//
// func main() {...}
//   The entry point of the program
//   Must be in package main
//   No parameters, no return value
//
// := operator
//   Short variable declaration
//   apiKey := "value" is same as: var apiKey string = "value"
//   Type is inferred from the value
//
// Error handling pattern:
//   if err != nil {
//       // handle error
//       return
//   }
//   Go doesn't have exceptions - errors are values
//   Always check errors!
//
// os.Getenv("VAR")
//   Gets environment variable (like $DEMETERICS_API_KEY in bash)
//   Returns empty string if not set
//
// os.Exit(1)
//   Exits program with status code
//   0 = success, non-zero = error
//
// json.Marshal(v)
//   Converts Go value to JSON bytes
//   Returns ([]byte, error)
//
// json.Unmarshal(data, &v)
//   Parses JSON bytes into Go value
//   &v = pointer to variable (so it can be modified)
//
// bytes.NewBuffer(data)
//   Creates an io.Reader from byte slice
//   Required for http.NewRequest body
//
// http.NewRequest(method, url, body)
//   Creates HTTP request
//   method = "GET", "POST", etc.
//   body can be nil for GET requests
//
// req.Header.Set(key, value)
//   Sets HTTP header
//   Use Add() to append multiple values
//
// client.Do(req)
//   Sends HTTP request
//   Returns response and error
//
// defer resp.Body.Close()
//   Schedules function call to run when current function returns
//   Ensures resources are cleaned up
//   Always defer Close() after checking error!
//
// io.ReadAll(reader)
//   Reads all data from io.Reader
//   Returns ([]byte, error)
//
// fmt.Printf("format", args...)
//   Formatted print
//   %v = default format
//   %d = decimal integer
//   %s = string
//   %f = float
//
// Capitalization matters:
//   apiKey = private (can't be used outside this package)
//   ApiKey = public (can be imported by other packages)
//
// Arrays vs Slices:
//   [5]int        = array, fixed size
//   []int         = slice, dynamic size (like arrays in other languages)
//   []Message{...} = slice literal
//
// Make vs literal:
//   make([]int, 5)     = create slice with length 5
//   []int{1, 2, 3}     = create slice from values
//   make(map[string]int) = create empty map

File: cpp/01_basic_chat.cpp
/**
 * ============================================================================
 * Example 1: Basic Chat - Your First AI API Call in C++
 * ============================================================================
 *
 * WHAT THIS DEMONSTRATES:
 *   - Making an HTTPS request to an AI API using libcurl
 *   - Building and parsing JSON with nlohmann/json (feels like Python!)
 *   - Modern C++ patterns: RAII, std::string, exceptions
 *   - Working with environment variables
 *
 * WHAT YOU'LL LEARN:
 *   - libcurl integration in C++
 *   - nlohmann/json intuitive syntax
 *   - C++ string handling (no manual memory management!)
 *   - Error handling with exceptions
 *
 * PREREQUISITES:
 *   - libcurl installed (see README.md)
 *   - nlohmann/json.hpp in lib/ folder or installed system-wide
 *   - DEMETERICS_API_KEY environment variable set
 *
 * COMPILE:
 *   g++ -std=c++17 -o 01_basic_chat 01_basic_chat.cpp -lcurl -I lib/
 *
 * RUN:
 *   ./01_basic_chat
 *
 * ============================================================================
 */

#include <iostream>     /* std::cout, std::cerr */
#include <string>       /* std::string */
#include <cstdlib>      /* getenv */
#include <curl/curl.h>  /* libcurl for HTTP requests */
#include "json.hpp"     /* nlohmann/json for JSON handling */

/* Use the nlohmann namespace for convenience */
using json = nlohmann::json;

/* ============================================================================
 * Step 1: Callback function for libcurl to write response data
 * ============================================================================
 * libcurl calls this function each time it receives data from the server.
 * We append the received data to a std::string (passed as userdata).
 *
 * Unlike C, we use std::string which manages its own memory!
 */
static size_t write_callback(void *contents, size_t size, size_t nmemb, void *userp) {
    size_t realsize = size * nmemb;
    std::string *response = static_cast<std::string*>(userp);
    response->append(static_cast<char*>(contents), realsize);
    return realsize;
}

/* ============================================================================
 * MAIN FUNCTION
 * ============================================================================
 */
int main() {
    /* Step 2: Get API key from environment variable */
    const char *api_key_cstr = std::getenv("DEMETERICS_API_KEY");
    if (api_key_cstr == nullptr) {
        std::cerr << "Error: DEMETERICS_API_KEY environment variable not set\n";
        std::cerr << "Run: export DEMETERICS_API_KEY='your_key_here'\n";
        return 1;
    }
    std::string api_key(api_key_cstr);

    /* Step 3: Initialize libcurl */
    curl_global_init(CURL_GLOBAL_ALL);
    CURL *curl = curl_easy_init();

    if (curl == nullptr) {
        std::cerr << "Error: Failed to initialize curl\n";
        curl_global_cleanup();
        return 1;
    }

    /* Step 4: Build the JSON request body using nlohmann/json */
    /* This syntax is incredibly intuitive - almost like Python! */
    json request_body = {
        {"model", "meta-llama/llama-4-scout-17b-16e-instruct"},
        {"messages", {
            {{"role", "user"}, {"content", "What is the capital of Switzerland?"}}
        }},
        {"temperature", 0.7},
        {"max_tokens", 100}
    };

    /* Convert JSON to string for HTTP request */
    std::string json_payload = request_body.dump();

    /* Step 5: Set up HTTP headers */
    struct curl_slist *headers = nullptr;
    headers = curl_slist_append(headers, "Content-Type: application/json");

    std::string auth_header = "Authorization: Bearer " + api_key;
    headers = curl_slist_append(headers, auth_header.c_str());

    /* Step 6: Initialize response string (no malloc needed in C++!) */
    std::string response_data;

    /* Step 7: Configure curl options */
    curl_easy_setopt(curl, CURLOPT_URL, "https://api.demeterics.com/groq/v1/chat/completions");
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, json_payload.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response_data);

    /* Step 8: Perform the HTTP request */
    std::cout << "Sending request to AI API...\n\n";
    CURLcode res = curl_easy_perform(curl);

    /* Step 9: Check for curl errors */
    if (res != CURLE_OK) {
        std::cerr << "Error: curl request failed: " << curl_easy_strerror(res) << "\n";
        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);
        curl_global_cleanup();
        return 1;
    }

    /* Step 10: Parse the JSON response */
    json response;
    try {
        response = json::parse(response_data);
    } catch (const json::parse_error &e) {
        std::cerr << "Error: Failed to parse response JSON: " << e.what() << "\n";
        std::cerr << "Raw response: " << response_data << "\n";
        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);
        curl_global_cleanup();
        return 1;
    }

    /* Step 11: Display the full response (pretty-printed) */
    std::cout << "Full Response:\n";
    std::cout << response.dump(2) << "\n";  /* 2 = indent spaces */

    /* Step 12: Extract and display the AI's answer */
    /* nlohmann/json makes this beautifully simple! */
    if (response.contains("choices") && !response["choices"].empty()) {
        std::string answer = response["choices"][0]["message"]["content"];
        std::cout << "\nAI Answer:\n" << answer << "\n";
    }

    /* Step 13: Display token usage */
    if (response.contains("usage")) {
        auto usage = response["usage"];
        std::cout << "\nToken Usage:\n";
        std::cout << "  Prompt: " << usage["prompt_tokens"] << "\n";
        std::cout << "  Response: " << usage["completion_tokens"] << "\n";
        std::cout << "  Total: " << usage["total_tokens"] << "\n";
    }

    /* Step 14: Clean up curl resources */
    /* Note: std::string and json clean themselves up automatically! (RAII) */
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);
    curl_global_cleanup();

    return 0;
}

/*
 * C++ CONCEPTS EXPLAINED:
 *
 * #include <iostream>
 *   C++ I/O library. Use std::cout for output, std::cin for input.
 *
 * std::string
 *   C++ string class. Manages its own memory - no malloc/free needed!
 *   Much safer than C-style char* strings.
 *
 * std::getenv("VAR")
 *   Same as C's getenv(), returns char* or nullptr.
 *
 * nullptr
 *   C++ null pointer (safer than NULL or 0).
 *
 * static_cast<Type*>(ptr)
 *   Safe type conversion between related pointer types.
 *   Checked at compile time.
 *
 * using json = nlohmann::json;
 *   Creates an alias for the nlohmann::json type.
 *   Like "typedef" but more powerful.
 *
 * json::parse(string)
 *   Parses a JSON string into a json object.
 *   Throws json::parse_error on failure.
 *
 * json.dump(indent)
 *   Converts json object to string.
 *   indent = pretty-print indentation (2 spaces).
 *
 * json["key"]
 *   Access JSON field by key (like Python dict).
 *   Throws exception if key doesn't exist.
 *
 * json.contains("key")
 *   Check if key exists without throwing.
 *
 * try { ... } catch (const Exception &e) { ... }
 *   Exception handling. Catches errors thrown by functions.
 *   e.what() returns error message.
 *
 * auto
 *   Compiler figures out the type automatically.
 *   auto x = 5;  // x is int
 *   auto s = "hello";  // s is const char*
 *
 * RAII (Resource Acquisition Is Initialization):
 *   Objects clean up after themselves when they go out of scope.
 *   std::string, std::vector, json all do this.
 *   No need for manual free() like in C!
 *
 * NLOHMANN/JSON SYNTAX:
 *   Creating JSON (feels like Python!):
 *     json j = {{"key", "value"}, {"number", 42}};
 *     json arr = {1, 2, 3, 4, 5};
 *     json nested = {{"user", {{"name", "Alice"}, {"age", 30}}}};
 *
 *   Reading JSON:
 *     std::string name = j["user"]["name"];
 *     int age = j["user"]["age"];
 *     for (auto& item : j["array"]) { ... }
 *
 * C++ vs C - KEY DIFFERENCES:
 *   1. std::string instead of char* (automatic memory)
 *   2. Exceptions instead of return codes (optional)
 *   3. RAII - objects clean up automatically
 *   4. << operator for output (std::cout << "hello")
 *   5. nullptr instead of NULL
 *   6. References (&) instead of just pointers (*)
 */
